### PR TITLE
#4337: SEGV Fault in the DES_fcrypt

### PR DESCRIPTION
**Pull request by ghost**
If these functions has to be compilant with crypt(), then \0 is also not allowed as a salt when salt contains \0 returned is NULL.

Changes in the line 74, 75. So when, there was salt[0] == \0, then value of the e_salt[1] was undefined, end then it was send to the ebcdic2ascii. (Hashtag Pure Evilness)

### CODE REVIEW

File: crypto/des/fcrypt.c [hunk 1]
```diff
@@ -70,9 +70,12 @@ char *DES_crypt(const char *buf, const char *salt)
     char e_buf[32 + 1];         /* replace 32 by 8 ? */
     char *ret;
 
+    if (sizeof(salt) < 2) return NULL;
```
Comment (ajmohan) [inline crypto/des/fcrypt.c]: sizeof(salt) is sizeof of the pointer. In this case, condition is always false.

Comment (richsalz) [inline crypto/des/fcrypt.c]: right; there's no way to test that the slt[] array is two bytes.  that's C for ya :)

File: test/destest.c [hunk 2]
```diff
@@ -718,6 +718,31 @@ int main(int argc, char *argv[])
         printf("fast crypt error, %s should be yA1Rp/1hZXIJk\n", str);
         err = 1;
     }
+    str = crypt("testing", "y€");
+    if (str != 0) {
```
Comment (richsalz) [inline test/destest.c]: please use ```NULL``` not zero for these tests.

**Comment by richsalz** [comment]
like the hashtag :)

File: crypto/des/fcrypt.c (hunk 1 above)
Comment (Ruzaki) [inline crypto/des/fcrypt.c]: Ups, I were testing sizeof against char salt[3]; and then it had sizeof(salt) == 3. My bad.
I wonder, commonly there are no NULL tests againts ptrs, in this one better write if (salt == NULL) return NULL; or just OpenSSLAssert()?

Comment (richsalz) [inline crypto/des/fcrypt.c]: Don’t do anything; if the caller passes in NULL the code will crash and they will fix things.

File: crypto/des/fcrypt.c [hunk 3]
```diff
@@ -70,9 +70,12 @@ char *DES_crypt(const char *buf, const char *salt)
...
+    if (salt[0] == '\0' || salt[1] == '\0') return NULL;
```
Comment (kroeckx) [inline crypto/des/fcrypt.c]: Why this change?  It seems 0 and 1 byte stings were valid before.

Comment (kroeckx) [inline crypto/des/fcrypt.c]: So that was based on someone else's comment

Comment (Ruzaki) [inline crypto/des/fcrypt.c]: It was "valid", with one size length array, but further else the salt[1] value was not set, so its content was not defined. And Linux crypt() function, requires that salt value is at least 2 char size array.

File: crypto/des/fcrypt.c [hunk 4]
```diff
@@ -70,9 +70,11 @@ char *DES_crypt(const char *buf, const char *salt)
     char e_buf[32 + 1];         /* replace 32 by 8 ? */
     char *ret;
 
+    if (salt[0] == '\0' || salt[1] == '\0') return NULL;
```
Comment (kroeckx) [inline crypto/des/fcrypt.c]: Can you move that return to the next line?

File: crypto/des/fcrypt.c [hunk 5]
```diff
@@ -115,17 +119,14 @@ char *DES_fcrypt(const char *buf, const char *salt, char *ret)
      * libraries.  People found that the disabled accounts effectively had no
      * passwd :-(.
      */
```
Comment (kroeckx) [inline crypto/des/fcrypt.c]: I'm wondering what to do with this comment now that we require the salt to be at least 2 bytes long.

**Comment by kroeckx** [comment]
DES_crypt() should check DES_fcrypt()'s return value before doing anything with it.

**Comment by kroeckx** [comment]
If we're going to force this 2 char salt minimum, we should probably also enforce that the characters are from the range of a-zA-Z0-9./, as documented in the crypt() manpage and POSIX, and we should document those things in the manual.  That probably also means `con_salt` can be changed.

**Comment by kroeckx** [comment]
So it seems the 2 char salt minimum was not in old versions of POSIX, but it's newer versions.  It used to say:

>  The first two characters of this string may be used to perturb the encoding algorithm. 

Now it says:

> The salt argument shall be a string of at least two bytes in length not including the null character

Also, it always said:

> The algorithm is implementation-dependent.

**Comment by kroeckx** [comment]
I'd also like to point out that with any change like this you might break someone that actually used a salt that's not 2 chars, and we should probably document that the old code used an `A` if the character wasn't used.

File: test/destest.c [hunk 6]
```diff
@@ -718,6 +718,31 @@ int main(int argc, char *argv[])
...
+    if (str != NULL) {
+        printf("salt error only usascii are accepted, returned value should be 0\n");
+        err = 1;
+    }
+    str = crypt("testing", "‰A");
+    if (str != NULL) {
+        printf("salt error only usascii are accepted, returned value should be 0\n");
+        err = 1;
+    }
+    str = crypt("testing", "\0A");
```
Comment (ajmohan) [inline test/destest.c]: The string is **'\', '0', 'A', '\0'**, so first two characters do not match nul ('\0').

**Comment by kroeckx** [comment]
The EBCDIC translation seems wrong to me.  If you use DES_crypt() the salt would get converted twice while if you use DES_fcrypt() the password would never get converted.

**Comment by mattcaswell** [comment]
Ping @ghost - there are a number of outstanding comments on this PR for you to respond to/resolve.

**Comment by mattcaswell** [comment]
Ping @ghost

**Comment by mattcaswell** [comment]
@kroeckx how should we proceed with this PR? No response from @ghost but there are outstanding issues with this PR. Should it be closed?

**Comment by kroeckx** [comment]
It looks like @ghost actually deleted his account.

**Comment by richsalz** [comment]
It's a real bug, either turn it into an RT or someone just fixes it.

**Comment by richsalz** [comment]
Replaced by #1147.
